torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -0,0 +1,638 @@
+        self._free_other_grads()
+
+        # Sync all the updated shards in between the ranks
+        self._broadcast_params()
+
+        # Sync hypothethical new results from the wrapped optimizer to the exposed param_groups
+        self._sync_param_groups(local_to_global=True)
+
+        return loss
+
+    def local_state_dict(self) -> Dict:
+        """Gets this rank's ``state_dict``.
+
+        Returns:
+            The state of the optimizer as a :class:`dict`.
+            It contains two entries:
+
+            * state - a dict holding current optimization state. Its content
+                differs between optimizer classes.
+            * param_groups - a dict containing all parameter groups
+        """
+        return self.optim.state_dict()
+
+    def consolidate_state_dict(self, recipient_rank: int = 0) -> None:
+        """Update the consolidated state_dict list, one per rank.
+
+        .. warning: This needs to be called on all replicas"""
+
+        # Sync lr and other attributes in case its been updated
+        self._sync_param_groups()
+
+        if self.rank == recipient_rank:
+            # Pull the sharded state from all the other replicas
+            # Store all the states in order, rank by rank
+            logging.debug("Pulling the sharded optimizer state from all replicas")
+            self._all_states = self._collect_sharded_states()
+        else:
+            # Acknowledge broadcasts, and send this rank's shard when needed
+            self._broadcast_state_dict()
+
+    def state_dict(self) -> Dict[str, Any]:
+        """
+        Returns:
+            the last known global optimizer state, which consist of a list of the shards.
+
+        .. warning:
+            If the state has not been consolidated, this returns a shard's worth, not the global state.
+
+        .. warning:
+            Returning the global state is limited to the replica which was responsible for the consolidation.
+            The state may also not be up to date, depending on when `consolidate_state_dict` was last called.
+        """
+
+        if len(self._all_states) == 0:
+            logging.warning("Optimizer state has not been consolidated. Returning the local state")
+            logging.warning("Please call `consolidate_state_dict()` beforehand if you meant to save the global state")
+            state_dict = self.local_state_dict()
+            state_dict["local_state_dict"] = True
+            return state_dict
+
+        # Flatten the param_groups, save the partition which logs the rank <> shard correspondence
+        partition: List[Tuple[int, int]] = []
+        param_groups: List[Dict[Any, Any]] = []
+
+        start = 0
+        for i, s in enumerate(self._all_states):
+            param_groups.extend(s["param_groups"])
+            end = start + len(s["param_groups"])
+            partition.append((start, end))
+            start = end
+
+        return {
+            "state": [s["state"] for s in self._all_states],
```

Comment:
Changing this in an update, thanks for the feedback, makes a lot of sense. I'm changing that to be completely compliant with the equivalent -non-sharded- pytorch optimizer, so that checkpoints become easy to mix and match, and elastic (could be reloaded with a different number of ranks). Getting there on https://github.com/facebookresearch/fairscale/pull/310

Reply:
discussed with @mrshenli , this will be fixed in a follow up PR in that we'll output a completely pytorch compliant state dict, as linked above